### PR TITLE
pyclass-impl: merge implementation traits

### PR DESCRIPTION
**Pull request by davidhewitt**
This PR merges `PyMethods`, `PyProtoMethods` and `PyClassSend` into a new trait `PyClassImpl`.

At the same time it does some small refactorings:
 - Replace `T::py_methods()` with `T::for_each_method_def` to avoid needing to make a temporary `Vec`
 - Moves `#[new]` and `#[call]` out of inventory to be supported internally by new traits & dtolnay specialization, because these are slot methods like `#[pyproto]`.
 - Removes `ctor` dependency (we don't need it any more).

Overall this is more a refactoring PR than anything particular. I plan to follow it up with another PR or two where I make inventory completely optional.

### CODE REVIEW

**Comment by davidhewitt** [comment]
To satisfy my curiosity I added a benchmark for type object initialization.

On master, on my laptop, it averages about 2900ns to initialize the the type object. With this branch, about 2700ns. So there's about 200ns of saving from this refactoring. Which will become meaningless for complex `#[pyclass]` types, and it's only one timie init, but still something!

**Comment by kngwyu** [comment]
Thanks!
I think the direction is good and will comment on some nitpicks when I find enough time.

**Comment by davidhewitt** [comment]
(rebased on master to fix merge conflicts)

File: src/class/impl_.rs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    /// This handles following two situations:
+    /// 1. In case `T` is `Send`, stub `ThreadChecker` is used and does nothing.
+    ///    This implementation is used by default. Compile fails if `T: !Send`.
+    /// 2. In case `T` is `!Send`, `ThreadChecker` panics when `T` is accessed by another thread.
+    ///    This implementation is used when `#[pyclass(unsendable)]` is given.
+    ///    Panicking makes it safe to expose `T: !Send` to the Python interpreter, where all objects
+    ///    can be accessed by multiple threads by `threading` module.
+    type ThreadChecker: PyClassThreadChecker<Self>;
+
+    fn for_each_method_def<Visitor: FnMut(&'static PyMethodDefType)>(_visitor: Visitor) {}
```
Comment (kngwyu) [inline src/class/impl_.rs:44]:
```suggestion
    fn for_each_method_def(_visitor: impl FnMut(&PyMethodDefType)) {}
```

File: src/class/impl_.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    fn get_new() -> ffi::newfunc {
+        fallback_new
+    }
+    fn get_call() -> Option<ffi::PyCFunctionWithKeywords> {
+        None
+    }
+    fn for_each_proto_slot<Visitor: FnMut(ffi::PyType_Slot)>(_visitor: Visitor) {}
```
Comment (kngwyu) [inline src/class/impl_.rs:51]:
```suggestion
    fn for_each_proto_slot(_visitor: impl FnMut(&ffi::PyType_Slot)) {}
```

File: pyo3-macros-backend/src/pyimpl.rs [hunk 3]
```diff
@@ -19,32 +20,46 @@ pub fn build_py_methods(ast: &mut syn::ItemImpl) -> syn::Result<TokenStream> {
...
+    let mut new_impls = Vec::new();
+    let mut call_impls = Vec::new();
     let mut methods = Vec::new();
-    let mut cfg_attributes = Vec::new();
     for iimpl in impls.iter_mut() {
         match iimpl {
             syn::ImplItem::Method(meth) => {
-                methods.push(pymethod::gen_py_method(ty, &mut meth.sig, &mut meth.attrs)?);
-                cfg_attributes.push(get_cfg_attributes(&meth.attrs));
+                match pymethod::gen_py_method(ty, &mut meth.sig, &mut meth.attrs)? {
```
Comment (kngwyu) [inline pyo3-macros-backend/src/pyimpl.rs:29]: This is really nice 👍🏼

File: pyo3-macros-backend/src/pyclass.rs [hunk 4]
```diff
@@ -419,41 +419,55 @@ fn impl_class(
...
+                let collector = PyClassImplCollector::<#cls>::new();
+                collector.new_impl()
+            }
+            fn get_call() -> Option<pyo3::ffi::PyCFunctionWithKeywords> {
+                use pyo3::class::impl_::*;
+                let collector = PyClassImplCollector::<#cls>::new();
+                collector.call_impl()
+            }
+
             fn for_each_proto_slot<Visitor: FnMut(pyo3::ffi::PyType_Slot)>(visitor: Visitor) {
```
Comment (kngwyu) [inline pyo3-macros-backend/src/pyclass.rs:446]:
```suggestion
            fn for_each_proto_slot(visitor: impl FnMut(&pyo3::ffi::PyType_Slot)) {
```

File: pyo3-macros-backend/src/pyclass.rs [hunk 5]
```diff
@@ -419,41 +419,55 @@ fn impl_class(
...
-                    .chain(protocols.buffer_protocol_slots())
+                    .chain(collector.number_protocol_slots())
+                    .chain(collector.iter_protocol_slots())
+                    .chain(collector.gc_protocol_slots())
+                    .chain(collector.descr_protocol_slots())
+                    .chain(collector.mapping_protocol_slots())
+                    .chain(collector.sequence_protocol_slots())
+                    .chain(collector.async_protocol_slots())
+                    .chain(collector.buffer_protocol_slots())
                     .cloned()
```
Comment (kngwyu) [inline pyo3-macros-backend/src/pyclass.rs:460]: Let's remove this and copy the slot later

File: pyo3-macros-backend/src/pyclass.rs [hunk 6]
```diff
@@ -419,41 +419,55 @@ fn impl_class(
...
-
         #into_pyobject
 
         #impl_inventory
 
-        impl pyo3::class::proto_methods::PyProtoMethods for #cls {
+        impl pyo3::class::impl_::PyClassImpl for #cls {
+            type ThreadChecker = #thread_checker;
+
+            fn for_each_method_def<Visitor: FnMut(&'static pyo3::class::PyMethodDefType)>(visitor: Visitor) {
```
Comment (kngwyu) [inline pyo3-macros-backend/src/pyclass.rs:429]:
```suggestion
            fn for_each_method_def(visitor: impl FnMut(&pyo3::class::PyMethodDefType)) {
```

**Review by kngwyu — APPROVED** [review]
Thanks for this hard work!
Left a few suggestions (mostly about impl trait).

**Comment by davidhewitt** [comment]
Thanks, somehow I missed that `impl Trait` is now allowed in argument position for trait methods! 😅

File: src/class/impl_.rs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    /// 1. In case `T` is `Send`, stub `ThreadChecker` is used and does nothing.
+    ///    This implementation is used by default. Compile fails if `T: !Send`.
+    /// 2. In case `T` is `!Send`, `ThreadChecker` panics when `T` is accessed by another thread.
+    ///    This implementation is used when `#[pyclass(unsendable)]` is given.
+    ///    Panicking makes it safe to expose `T: !Send` to the Python interpreter, where all objects
+    ///    can be accessed by multiple threads by `threading` module.
+    type ThreadChecker: PyClassThreadChecker<Self>;
+
+    fn for_each_method_def(_visitor: impl FnMut(&PyMethodDefType)) {}
+    fn get_new() -> ffi::newfunc {
```
Comment (kngwyu) [inline src/class/impl_.rs:45]: How about making the return type `Option<ffi::newfunc>` and use `unwrap_or(fallback_new)` in the caller?
Just a minor detail, though.

Comment (davidhewitt) [reply src/class/impl_.rs:45]: Agreed, I think that's more consistent with `get_call` in that way.

**Comment by kngwyu** [comment]
Thanks!

**Pull request merged**